File: zig/day02/submarine.zig
const std = @import("std");
const Move = @import("move.zig").Move;
const Self = @This();

h_pos: i32,
depth: i32,
aim: i32,

pub fn part1(moves: []Move) i32 {
  var self = std.mem.zeroes(Self);
  for (moves) |move| {
    switch (move) {
      .down => |n| self.depth += n,
      .forward => |n| self.h_pos += n,
    }
  }
  return self.h_pos * self.depth;
}

pub fn part2(moves: []Move) i32 {
  var self = std.mem.zeroes(Self);
  for (moves) |move| {
    switch (move) {
      .down => |n| self.aim += n,
      .forward => |n| {
        self.h_pos += n;
        self.depth += n * self.aim;
      }
    }
  }
  return self.h_pos * self.depth;
}

File: kotlin/day02/Submarine.kt
package aoc.day02

internal class Submarine {
  private var hPos = 0
  private var depth = 0
  private var aim = 0

  fun part1(moves: Iterable<Move>): Int {
    for (move in moves) when (move) {
      is Move.Down -> depth += move.steps
      is Move.Forward -> hPos += move.steps
    }
    return hPos * depth
  }

  fun part2(moves: Iterable<Move>): Int {
    for (move in moves) when (move) {
      is Move.Down -> aim += move.steps
      is Move.Forward -> {
        hPos += move.steps
        depth += move.steps * aim
      }
    }
    return hPos * depth
  }
}